File: elixir/lib/bonemesh/chunk.ex
defmodule Bonemesh.Chunk do
  @moduledoc """
  Splitting of oversized application payloads (protocol.md §6.1).

  A payload too large for one transport frame is serialized, its UTF-8 bytes cut
  into segments of at most `max_segment_bytes/0` on character boundaries, and each
  segment sent as a data message sharing one message id, carrying `chunk {i, n}`
  and a top-level `"seg"` string and **no** `"payload"`. A payload that fits
  travels whole, with `"payload"` and no `"seg"`.

  Segments are text, not Base64: §0's Base64 rule covers binary fields, a slice of
  JSON text is already UTF-8, and a JSON string carries it directly, so a split
  message stays readable through the key-log inspector (decisions #3, #5, #25).
  Cutting on a byte budget rather than a character count is what keeps the split
  identical across the seven implementations, because UTF-8 has no surrogates: a
  code point is either wholly inside a segment or wholly outside it. An Elixir
  binary is already a byte sequence, so `binary-size` is the right primitive here
  and `String.slice/3`, which counts graphemes, would be the wrong one.

  Reassembly is `Bonemesh.Reassembler`.
  """

  @max_segment_bytes 24000
  @max_chunks 1024
  @max_reassembly_buffer 16_777_216
  @max_concurrent_reassemblies 256
  @reassembly_timeout_millis 30_000

  @doc "Maximum payload bytes carried by one segment (protocol.md §0)."
  def max_segment_bytes, do: @max_segment_bytes

  @doc "Maximum segments one application message may be split into (§0)."
  def max_chunks, do: @max_chunks

  @doc "Maximum segment bytes buffered at once, across every in-flight message (§0)."
  def max_reassembly_buffer, do: @max_reassembly_buffer

  @doc "Maximum messages that may be mid-reassembly at once (§0)."
  def max_concurrent_reassemblies, do: @max_concurrent_reassemblies

  @doc "Milliseconds a partially-filled message may sit before being discarded (§0)."
  def reassembly_timeout_millis, do: @reassembly_timeout_millis

  @doc """
  Splits a payload into one whole data message or a series of segments.

  Raises `ArgumentError` when no conforming destination would reassemble it, so the
  caller is told locally rather than the mesh carrying a message that cannot arrive
  (§6.1, Bounds).
  """
  def split(mid, from, to, ttl, payload) do
    src = JSON.encode!(payload)

    cond do
      byte_size(src) <= @max_segment_bytes ->
        [Bonemesh.Message.data(mid, from, to, ttl, payload)]

      byte_size(src) > @max_reassembly_buffer ->
        raise ArgumentError,
              "payload of #{byte_size(src)} bytes exceeds the reassembly buffer maximum " <>
                "of #{@max_reassembly_buffer}"

      true ->
        segments = cut(src)
        n = length(segments)

        if n > @max_chunks do
          raise ArgumentError, "payload needs #{n} segments, over the maximum of #{@max_chunks}"
        end

        segments
        |> Enum.with_index()
        |> Enum.map(fn {seg, i} ->
          Bonemesh.Message.data_segment(mid, from, to, ttl, i, n, seg)
        end)
    end
  end

  # Cuts a binary into at-most-@max_segment_bytes pieces, every cut landing on a
  # character boundary so each piece is itself valid UTF-8.
  defp cut(<<>>), do: []

  defp cut(s) when byte_size(s) <= @max_segment_bytes, do: [s]

  defp cut(s) do
    at = char_boundary(s, @max_segment_bytes)
    <<head::binary-size(at), rest::binary>> = s
    [head | cut(rest)]
  end

  # Walks a proposed cut back to the nearest character boundary at or before it.
  defp char_boundary(s, at) do
    case :binary.at(s, at) do
      b when Bitwise.band(b, 0xC0) == 0x80 and at > 1 -> char_boundary(s, at - 1)
      _ -> at
    end
  end
end

File: python/bonemesh/chunk.py
"""Splitting and reassembly of oversized application payloads (protocol.md section 6.1).

A payload too large for one transport frame is serialized, its UTF-8 bytes cut into
segments of at most MAX_SEGMENT_BYTES on character boundaries, and each segment sent
as a data message sharing one message id, carrying ``chunk {i, n}`` and a top-level
``seg`` string and **no** ``payload``. A payload that fits travels whole, with
``payload`` and no ``seg``.

Segments are text, not Base64: section 0's Base64 rule covers binary fields, a slice
of JSON text is already UTF-8, and a JSON string carries it directly, so a split
message stays readable through the key-log inspector (decisions #3, #5, #25).
Cutting on a byte budget rather than a character count is what keeps the split
identical across the seven implementations, because UTF-8 has no surrogates: a code
point is either wholly inside a segment or wholly outside it. This port therefore
encodes to bytes first and slices those, rather than slicing the str -- a str index
is a code point, which would produce different cuts from every other port.
"""

from __future__ import annotations

import json

from . import message

#: Maximum payload bytes carried by one segment (protocol.md section 0).
MAX_SEGMENT_BYTES = 24000
#: Maximum segments one application message may be split into (section 0).
MAX_CHUNKS = 1024
#: Maximum segment bytes buffered at once, across every in-flight message (section 0).
MAX_REASSEMBLY_BUFFER = 16777216
#: Maximum messages that may be mid-reassembly at once (section 0).
MAX_CONCURRENT_REASSEMBLIES = 256
#: Milliseconds a partially-filled message may sit before being discarded (section 0).
REASSEMBLY_TIMEOUT_MILLIS = 30000


def _char_boundary(src: bytes, start: int, end: int) -> int:
    """Walk a proposed cut back to the nearest character boundary at or before it,
    so a segment never ends mid-character and is itself valid UTF-8."""
    if end >= len(src):
        return end  # the tail is always a boundary
    e = end
    while e > start and src[e] & 0xC0 == 0x80:  # 10xxxxxx is a continuation byte
        e -= 1
    # A UTF-8 character is at most 4 bytes and a segment is 24000, so e cannot reach
    # start from well-formed input; falling back keeps a malformed serializer from
    # producing a zero-length segment and looping forever.
    return e if e > start else end


def split(mid: str, frm: str, to: str, ttl: int, payload) -> list[dict]:
    """Split a payload into one whole data message or a series of segments.

    Raises ``ValueError`` when no conforming destination would reassemble it, so the
    caller is told locally rather than the mesh carrying a message that cannot arrive
    (section 6.1, Bounds).
    """
    src = json.dumps(payload, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
    if len(src) <= MAX_SEGMENT_BYTES:
        return [message.data(mid, frm, to, ttl, payload)]
    if len(src) > MAX_REASSEMBLY_BUFFER:
        raise ValueError(
            f"payload of {len(src)} bytes exceeds the reassembly buffer maximum "
            f"of {MAX_REASSEMBLY_BUFFER}"
        )

    segs: list[str] = []
    pos = 0
    while pos < len(src):
        end = _char_boundary(src, pos, min(pos + MAX_SEGMENT_BYTES, len(src)))
        segs.append(src[pos:end].decode("utf-8"))
        pos = end
    if len(segs) > MAX_CHUNKS:
        raise ValueError(f"payload needs {len(segs)} segments, over the maximum of {MAX_CHUNKS}")

    n = len(segs)
    return [message.data_segment(mid, frm, to, ttl, i, n, seg) for i, seg in enumerate(segs)]


class _Partial:
    __slots__ = ("segments", "received", "nbytes", "started")

    def __init__(self, n: int, started: int) -> None:
        self.segments: list[str | None] = [None] * n
        self.received = 0
        self.nbytes = 0
        self.started = started


class Reassembler:
    """Rebuild split payloads at the destination, the counterpart to :func:`split`.

    Every section 0 bound is enforced **before** any allocation keyed on a number the
    peer chose. That ordering is the point: the Java reference sized its buffer on
    the peer's ``n`` and validated afterwards, so one frame claiming two billion
    segments exhausted the heap -- defect D7 reintroduced by the feature meant to fix
    it.

    Three bounds, none redundant. The byte budget caps one large message; the
    in-flight count caps a flood of distinct ids each carrying an *empty* segment,
    which costs nothing against a byte budget and still costs memory; the timeout
    stops an abandoned message pinning memory for the session's life.
    """

    #: sentinel distinguishing "nothing to deliver" from "the payload is None",
    #: which a bare None return could not
    INCOMPLETE = object()

    def __init__(self) -> None:
        # dicts keep insertion order, which is what lets the sweep stop early
        self._partials: dict[str, _Partial] = {}
        self._buffered = 0

    def offer(self, msg: dict, now_millis: int):
        """Feed one inbound data message.

        Returns the payload when a message completes, else :data:`INCOMPLETE`.
        ``now_millis`` is passed in rather than read so the timeout is testable
        without sleeping.
        """
        self._sweep(now_millis)

        has_chunk = "chunk" in msg
        chunk = msg.get("chunk")
        if has_chunk and not isinstance(chunk, dict):
            return self.INCOMPLETE  # chunk is present but not an object
        n = 1 if not has_chunk else (chunk.get("n") if message._is_int(chunk.get("n")) else None)
        if has_chunk and n is None:
            return self.INCOMPLETE

        if not has_chunk or n == 1:
            # A whole message carries payload and no seg. One claiming n == 1 while
            # carrying seg instead is malformed, not a one-segment split -- and so is
            # one carrying both.
            if "seg" in msg:
                return self.INCOMPLETE
            return msg["payload"] if "payload" in msg else self.INCOMPLETE

        # Bounds first, allocation second.
        i = chunk.get("i") if message._is_int(chunk.get("i")) else None
        if i is None or n < 1 or n > MAX_CHUNKS or i < 0 or i >= n:
            return self.INCOMPLETE
        seg = msg.get("seg")
        if not isinstance(seg, str):
            return self.INCOMPLETE  # a segment without its slice
        mid = msg.get("mid")
        if not isinstance(mid, str):
            return self.INCOMPLETE

        partial = self._partials.get(mid)
        if partial is None:
            if len(self._partials) >= MAX_CONCURRENT_REASSEMBLIES:
                return self.INCOMPLETE
            partial = _Partial(n, now_millis)
            self._partials[mid] = partial
        elif len(partial.segments) != n:
            self._discard(mid)  # the peer changed n mid-message
            return self.INCOMPLETE

        if partial.segments[i] is None:
            size = len(seg.encode("utf-8"))
            if self._buffered + size > MAX_REASSEMBLY_BUFFER:
                self._discard(mid)
                return self.INCOMPLETE
            partial.segments[i] = seg
            partial.nbytes += size
            partial.received += 1
            self._buffered += size
        if partial.received != n:
            return self.INCOMPLETE

        joined = "".join(partial.segments)  # type: ignore[arg-type]
        self._discard(mid)
        try:
            return json.loads(joined)
        except ValueError:
            return self.INCOMPLETE  # the segments did not rebuild valid JSON

    def in_flight(self) -> int:
        """Messages currently mid-reassembly.

        Exposed for the tests that assert the bounds release memory rather than
        merely refusing to add to it -- a reassembler that rejects a segment but
        keeps its partial forever is still a leak, and the refusal alone cannot show
        that.
        """
        return len(self._partials)

    def buffered(self) -> int:
        """Segment bytes held across every in-flight message."""
        return self._buffered

    def _discard(self, mid: str) -> None:
        partial = self._partials.pop(mid, None)
        if partial is not None:
            self._buffered -= partial.nbytes

    def _sweep(self, now_millis: int) -> None:
        for mid in list(self._partials):
            partial = self._partials[mid]
            if now_millis - partial.started < REASSEMBLY_TIMEOUT_MILLIS:
                break  # insertion-ordered: the rest are younger
            self._buffered -= partial.nbytes
            del self._partials[mid]
